pkg/adaptor/cloud/util.go

```diff
@@ -18,3 +24,89 @@ func DefaultToEnv(field *string, env, fallback string) {
 
 	*field = val
 }
+
+// Method to verify the correct instanceType to be used for Pod VM
+func VerifyCloudInstanceType(instanceType string, validInstanceTypes []string, defaultInstanceType string) (string, error) {
+	// If instanceType is empty, set instanceType to default.
+	if instanceType == "" {
+		instanceType = defaultInstanceType
+		fmt.Printf("Using default instance type (%q)\n", defaultInstanceType)
+		return instanceType, nil
+	}
+
+	// If instanceTypes is empty and instanceType is not default, return error
+	if len(validInstanceTypes) == 0 && instanceType != defaultInstanceType {
+		// Return error if instanceTypes is empty and instanceType is not default
+		return "", fmt.Errorf("requested instance type (%q) is not default (%q) and supported instance types list is empty",
+			instanceType, defaultInstanceType)
+
+	}
+
+	// If instanceTypes is not empty and instanceType is not among the supported instance types, return error
+	if len(validInstanceTypes) > 0 && !util.Contains(validInstanceTypes, instanceType) {
+		return "", fmt.Errorf("requested instance type (%q) is not part of supported instance types list", instanceType)
+	}
+
+	return instanceType, nil
+}
+
+// Method to sort InstanceTypeSpec into ascending order based on memory
+func SortInstanceTypesOnMemory(instanceTypeSpecList []InstanceTypeSpec) []InstanceTypeSpec {
+
+	// Use sort.Slice to sort the instanceTypeTupleList slice and return the sorted slice
+	sort.Slice(instanceTypeSpecList, func(i, j int) bool {
+		return instanceTypeSpecList[i].Memory < instanceTypeSpecList[j].Memory
+	})
+
+	return instanceTypeSpecList
+}
+
+func SelectInstanceTypeToUse(spec InstanceTypeSpec, specList []InstanceTypeSpec, validInstanceTypes []string, defaultInstanceType string) (string, error) {
+
+	var instanceType string
+	var err error
+
+	// If vCPU and memory are set in annotations then find the best fit instance type
+	// from the cloud provider
+	// vCPU and Memory gets higher priority than instance type from annotation
+	if spec.VCPUs != 0 && spec.Memory != 0 {
+		instanceType, err = GetBestFitInstanceType(specList, spec.VCPUs, spec.Memory)
+		if err != nil {
+			return "", fmt.Errorf("failed to get instance type based on vCPU and memory annotations: %w", err)
+		}
+		logger.Printf("Instance type selected by the cloud provider based on vCPU and memory annotations: %s", instanceType)
+	} else if spec.InstanceType != "" {
+		instanceType = spec.InstanceType
+		logger.Printf("Instance type selected by the cloud provider based on instance type annotation: %s", instanceType)
+	}
```

Comment:
Do vcpus and memory have a higher precedence than instance type (machine type annotation)?

I thought the instance type annotation has a higher precedence, since it means that a user explicitly sets a specific instance type in a pod definition. I also think that we can make these annotations exclusive options, and causes an error when both annotations are specified.

@bpradipt wdyt?

Reply:
@yoheiueda yes vcpus and mem have higher precedence than the annotation. I added that as a comment and will also include it in the docs. It's debatable thought. My reasoning has been that if the user is taking the initiative to specify vcpus and memory then it likely indicates a need for best fit and should take precedence. 
I don't think we need to make the options mutually exclusive. Precedence will provide flexibility and ease of use during deployment.

Reply:
OK, I see.